---
title: "The *exomePeak2* User's Guide"
author: 
- name: Zhen Wei
  affiliation:
  - Department of Biological Sciences, Xi’an Jiaotong-Liverpool University, Suzhou, Jiangsu, 215123, China
  - Institute of Integrative Biology, University of Liverpool, L7 8TX, Liverpool, United Kingdom
  email: Zhen.Wei01@xjtlu.edu.cn
- name: Jia Meng
  affiliation:
  - Department of Biological Sciences, Xi’an Jiaotong-Liverpool University, Suzhou, Jiangsu, 215123, China
  - Institute of Integrative Biology, University of Liverpool, L7 8TX, Liverpool, United Kingdom
  email: Jia.Meng@xjtlu.edu.cn
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
graphics: yes
vignette: >
  %\VignetteIndexEntry{The exomePeak2 user's guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignettePackage{exomePeak2}
  %\VignetteEncoding{UTF-8}
---

```{r para, echo = FALSE, results='hide'}
knitr::opts_chunk$set(dev="png",fig.show="hold",
               fig.width=8,fig.height=4.5,fig.align="center",
               message=FALSE,collapse=TRUE)
```

# Introduction

**exomePeak2** provides bias-aware quantification and peak detection for Methylated RNA immunoprecipitation sequencing data (**MeRIP-Seq**). *MeRIP-Seq* is a commonly applied sequencing technology that can measure the location and abundance of RNA modification sites under given cell line conditions. However, quantification and peak calling in *MeRIP-Seq* are sensitive to PCR amplification biases, which generally present in next-generation sequencing (NGS) technologies. In addition, the count data generated by **RNA-Seq** exhibits significant biological variations between biological replicates. *exomePeak2* collectively address the challenges by introducing a series of robust data science tools tailored for MeRIP-Seq. Using *exomePeak2*, users can perform peak calling, modification site quantification and differential analysis through a straightforward single-step function. Alternatively, multi-step functions can be used to generate diagnostic plots and perform customized analyses.

# Installation

To install exomePeak2 from bioconductor, start R (version >"3.6") and enter:

```{r, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("exomePeak2")
```

For order versions of R, please refer to the appropriate [Bioconductor release](https://www.bioconductor.org/about/release-announcements/}.

# Peak Calling

For the peak calling of the *MeRIP-Seq* experiment, exomePeak2 requires that the alignment result be stored in **BAM** format. User can specify the corresponding **BAM** directories of IP and input samples at the arguments `bam_ip` and `bam_input` respectively. 

The following example demonstrates the peak calling from **BAM** and **GFF** files.  Besides **GFF** file, transcript annotation can also be provided by the bioconductor **TxDb** object. The **TxDb** will be automatically downloaded if the corresponding UCSC genome name is filled in the `genome` argument.

The genome sequence is necessary for the correction of GC content bias. If both the `genome` and the `bsgenome` arguments are missing ( `= NULL` ), exomPeak2 will perform peak calling without correcting PCR amplification bias.

```{r, eval = TRUE}
library(exomePeak2)

set.seed(1)

GENE_ANNO_GTF = system.file("extdata", "example.gtf", package="exomePeak2")

f1 = system.file("extdata", "IP1.bam", package="exomePeak2")
f2 = system.file("extdata", "IP2.bam", package="exomePeak2")
f3 = system.file("extdata", "IP3.bam", package="exomePeak2")
f4 = system.file("extdata", "IP4.bam", package="exomePeak2")
IP_BAM = c(f1,f2,f3,f4)

f1 = system.file("extdata", "Input1.bam", package="exomePeak2")
f2 = system.file("extdata", "Input2.bam", package="exomePeak2")
f3 = system.file("extdata", "Input3.bam", package="exomePeak2")
INPUT_BAM = c(f1,f2,f3)

exomePeak2(bam_ip = IP_BAM,
           bam_input = INPUT_BAM,
           gff = GENE_ANNO_GTF,
           genome = "hg19")
```

exomePeak2 will export the significant peaks in the format of **BED** file and **CSV** table, and the output results will be automatically saved in a folder named `exomePeak2_output`.

If only IP and input samples are provided, the peak statistics are calculated from the ${\beta}_{i,1}$ terms in the following linear regression design:

$$log2(Q_{i,j}) = {\beta}_{i,0} + {\beta}_{i,1}1(\rho(j)=\text{IP}) + t_{i,j}$$

$Q_{i,j}$ is the expected value of read abundance of peak $i$ under sample $j$. ${\beta}_{i,0}$ is the intercept coefficient, ${\beta}_{i,1}$ is the coefficient of IP/input log2 fold change, $1(\rho(j)=\text{IP})$ is the regression covariate, which is the indicator variable of the sample $j$ being IP sample. $t_{i,j}$ is the peak specific offsets account for the sequencing depth variation and the GC content biases. $t{i,j}$ is estimated by the product of median read count on background bins (classified by GMM) and centered Poisson GLM fits on read count with fragment GC content as covariate. Knots for smoothing splines used in Poisson GLM are set at c(0, 0.4, 0.5, 0.6, 1).

For peak calling, Poisson GLMs using the design above are fitted to identify the significantly modified sliding windows; the detection criteria is Wald test p-values < 1e-05 under the alternative hypothesis of ${\beta}_{i,1} > 0$. 

Annotations on the columns of the output table:

- ***chr***: the chromosome name of the peak.
- ***chromStart***: the start of the peak on the chromosome.
- ***chromEnd***: the end of the peak on the chromosome.
- ***name***: the unique ID of the modification peak.
- ***strand***: the strand of the peak on genome.
- ***blockCount***: the block (exon) number within the peak.
- ***blockSizes***: the widths of the blocks.
- ***blockStarts***: the start positions of the blocks.
- ***geneID***: the gene ID of the peak.
- ***RPM.input***: the RPM (Reads per million mapped reads) calculated from the pooled input samples.
- ***RPM.IP***: the RPM calculated from the pooled IP samples.
- ***log2FC***: the estimate of IP over input log2 fold change (coefficient estimates of ${\beta}_{i,1}$).
- ***pvalue***: the p-value of the peak, which is calculated by the Wald test over the coefficient of ${\beta}_{i,1}$.
- ***fdr***: the adjusted Wald test p-value using Benjamini Hochberg approach.
- ***score***: the -log2 p value of the peak.

# Differential Modification Analysis 

The following example performs differential modification analysis (comparison of two biological conditions) on the exon regions defined by the transcript annotation.

In the differential modification mode, exomePeak2 will first perform Peak calling on the exon regions using both the control and treated samples. Next, it will conduct the differential modification analysis over the read counts of significant peaks using GLMs of interactive designs.

```{r, eval = TRUE}
f1 = system.file("extdata", "treated_IP1.bam", package="exomePeak2")
TREATED_IP_BAM = c(f1)
f1 = system.file("extdata", "treated_Input1.bam", package="exomePeak2")
TREATED_INPUT_BAM = c(f1)

exomePeak2(bam_ip = IP_BAM,
           bam_input = INPUT_BAM,
           bam_input_treated = TREATED_INPUT_BAM,
           bam_ip_treated = TREATED_IP_BAM,
           gff = GENE_ANNO_GTF,
           genome = "hg19")
```

In the differential modification mode, exomePeak2 will export the differentially modified peaks in the format of **BED** file and **CSV** table, and the outputs will also be saved under the folder `exomePeak2_output`.

The peak statistics calculated in the differential modification setting are based on the interactive coefficient ${\beta}_{i,3}$ in the following regression design:

$$log2(Q_{i,j}) = {\beta}_{i,0} + {\beta}_{i,1}1(\rho(j)=\text{IP}) + {\beta}_{i,2}1(\rho(j)=\text{Treatment}) + {\beta}_{i,3}1(\rho(j)=\text{IP&Treatment}) + t_{i,j}$$

The ***diff.log2FC*** and ***pvalue*** in the output of differential analysis are calculated over the coefficient estimates of the interactive term ${\beta}_{i,3}$. The differential log2FC here can be interpreted as the log(modification FC in treated group / modification FC in control group).

# Contact

If you encounter any problems during use, please contact the maintainer of exomePeak2:

**Zhen Wei** :  <zhen.wei01@xjtlu.edu.cn>

# References
1. KD Hansen, RA Irizarry, and Z Wu, Removing technical variability in RNA-seq data using conditional quantile normalization. Biostatistics 2012 vol. 13(2) pp. 204-216.

2. Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15:550. https://doi.org/10.1186/s13059-014-0550-8

3. Zhu A, Ibrahim JG, Love MI (2018). “Heavy-tailed prior distributions for sequence count data: removing the noise and preserving large differences.” Bioinformatics. doi: 10.1093/bioinformatics/bty895.

# Session Info
```{r}
sessionInfo()
```


